fdb-extensions/src/main/java/com/apple/foundationdb/async/RankedSet.java

```diff
@@ -220,32 +391,110 @@ public RankedSet(Subspace subspace, Executor executor) {
+                            }
+                            // This works even if the current set does not track duplicates but duplicates were added
+                            // earlier by one that did.
+                            final boolean duplicate = count > 1;
+                            final int nlevels = config.getNLevels();
+                            final List<CompletableFuture<Void>> futures = new ArrayList<>(nlevels);
+                            for (int li = 0; li < nlevels; ++li) {
+                                final int level = li;
+
+                                final CompletableFuture<Void> future;
+
+                                if (duplicate) {
+                                    // Always subtract one, never clearing a level count key.
+                                    // Concurrent requests both subtracting one when the count is two will conflict
+                                    // on the level zero key, which was read to detect that this is a duplicate.
+                                    // So it should not be possible for a count to go to zero.
+                                    Function<byte[], Void> decrement = k -> {
+                                        tr.mutate(MutationType.ADD, subspace.pack(Tuple.from(level, k)), encodeLong(-1));
+                                        return null;
+                                    };
+                                    if (level == 0) {
+                                        decrement.apply(key);
+                                        future = DONE;
+                                    } else {
+                                        future = getPreviousKey(tr, level, key, true).thenApply(decrement);
+                                    }
+                                } else {
+                                    // This could be optimized to check the hash for which levels should have this key.
+                                    // That would require that the hash function never changes, though.
+                                    // This allows for it to change, with the distribution perhaps getting a little uneven
+                                    // as a result. It even allows for the hash function to return a random number.
+                                    // It also further guarantees that counts never go to zero.
+                                    final byte[] k = subspace.pack(Tuple.from(level, key));
+                                    final CompletableFuture<byte[]> cf = tr.get(k);
+
+                                    if (level == 0) {
+                                        future = cf.thenApply(c -> {
+                                            if (c != null) {
+                                                tr.clear(k);
+                                            }
+                                            return null;
+                                        });
```

Comment:
It seems weird (though perhaps) harmless that this re-reads the key for level zero rather than, say, just clearing the key. For one, it almost gives the impression that the key could be null at this point...which seems wrong.

Reply:
Yes, it is simpler to just clear unconditionally.